### DISCUSSION TITLE
Implement PgRangeExpressionMethods for a custom range

### DISCUSSION THREAD

**Discussion by booti386** [diesel-rs/diesel#3875] [Q&A] [answered]
Follow-up of https://github.com/diesel-rs/diesel/discussions/3870

To be able to use .contains() on a custom range (`Timerange`), I need to implement `PgRangeExpressionMethods` for it.

I tried:
```rust
pub trait TimerangeOrNullableTimerange {}

impl TimerangeOrNullableTimerange for Timerange {}
impl TimerangeOrNullableTimerange for Nullable<Timerange> {}

impl<T> PgRangeExpressionMethods for T
where
    T: Expression,
    T::SqlType: TimerangeOrNullableTimerange,
{}
```

Which obviously fails because `PgRangeExpressionMethods` is not defined in the current crate.

Is also could have implemented `RangeOrNullableRange` on `Timerange` and `Nullable<Timerange>`, however it is a private trait so i can't.

What is the right/best way to make the .contains() method available to this type?

**Comment by weiznich** [comment, marked as answer]
As you already noticed it's not possible to use the existing traits for this. You need to define your own custom operator via [`diesel::infix_operator!`](https://docs.diesel.rs/2.1.x/diesel/macro.infix_operator.html) and provide a custom extension trait on your own. The linked documentation as a generic example for this.

**Reply by booti386** [reply]
Okay, than you for the quick reply.
